spring-graphql/src/main/java/org/springframework/graphql/data/QuerydslDataFetcherSupport.java

```diff
@@ -0,0 +1,260 @@
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+
+import org.springframework.core.ResolvableType;
+import org.springframework.core.annotation.MergedAnnotations;
+import org.springframework.core.convert.support.DefaultConversionService;
+import org.springframework.data.mapping.model.EntityInstantiators;
+import org.springframework.data.projection.ProjectionFactory;
+import org.springframework.data.projection.SpelAwareProxyProjectionFactory;
+import org.springframework.data.querydsl.QuerydslPredicateExecutor;
+import org.springframework.data.querydsl.SimpleEntityPathResolver;
+import org.springframework.data.querydsl.binding.QuerydslBinderCustomizer;
+import org.springframework.data.querydsl.binding.QuerydslBindings;
+import org.springframework.data.querydsl.binding.QuerydslPredicateBuilder;
+import org.springframework.data.repository.NoRepositoryBean;
+import org.springframework.data.repository.Repository;
+import org.springframework.data.repository.core.support.DefaultRepositoryMetadata;
+import org.springframework.data.util.ClassTypeInformation;
+import org.springframework.data.util.Streamable;
+import org.springframework.data.util.TypeInformation;
+import org.springframework.util.Assert;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+
+/**
+ * Utility to implement {@link DataFetcher} based on Querydsl {@link Predicate} through {@link QuerydslPredicateExecutor}.
+ * Actual instances can be created through a {@link #builder(QuerydslPredicateExecutor) builder} to query for
+ * {@link Builder#single()} or {@link Builder#many()} objects.
+ * Example:
+ * <pre class="code">
+ * interface BookRepository extends Repository&lt;Book, String&gt;, QuerydslPredicateExecutor&lt;Book&gt;{}
+ *
+ * BookRepository repository = …;
+ * TypeRuntimeWiring wiring = …;
+ *
+ * wiring.dataFetcher("books", QuerydslDataFetcherSupport.builder(repository).many())
+ *       .dataFetcher("book", QuerydslDataFetcherSupport.builder(repositories).single());
+ * </pre>
+ *
+ * @param <T> returned result type
+ * @author Mark Paluch
+ * @since 1.0.0
+ * @see QuerydslPredicateExecutor
+ * @see Predicate
+ * @see QuerydslBinderCustomizer
+ */
+public abstract class QuerydslDataFetcherSupport<T> {
+
+	private static final QuerydslPredicateBuilder BUILDER = new QuerydslPredicateBuilder(DefaultConversionService
+			.getSharedInstance(), SimpleEntityPathResolver.INSTANCE);
+
+	private final TypeInformation<?> domainType;
+
+	private final QuerydslBinderCustomizer<EntityPath<?>> customizer;
+
+	QuerydslDataFetcherSupport(QuerydslPredicateExecutor<T> executor, QuerydslBinderCustomizer<EntityPath<?>> customizer) {
+
+		this.customizer = customizer;
+
+		Class<?> repositoryInterface = getRepositoryInterface(executor);
+
+		DefaultRepositoryMetadata metadata = new DefaultRepositoryMetadata(repositoryInterface);
+		domainType = ClassTypeInformation.from(metadata.getDomainType());
```

Comment:
I wonder if the work done in the constructor could be done earlier, in the builder, and simply passed in. That way the support class becomes a simple holder of the shared fields.

Reply:
Makes sense to catch any exceptions as early as possible.